### PR TITLE
Test/shef parser

### PR DESCRIPTION
**Pull request by ktarbet**
Exploring Fix for #384

## Solution

Attempting to create a shef parsing test using the test/java/fixtures/DecodesHelper. 

I've made some modifications to DecodesHelper, and made some progress.  now the ShefProcess is confused because the SensorNumber = -1, so data is being dropped.

Considering if I should use integration test instead...

### CODE REVIEW

**Comment by ktarbet** [comment]
@MikeNeilson Do you know the design for the .sensors testing input file?  I see a  ```TODO: lookup algo and parts``` in DecodesHelper  , and I'm not sure what the purpose of the ';' semi-colon is.

I would like to include DataType information in the .sensors file.  I've hacked an experiment in commit [5c4d2e1](https://github.com/opendcs/opendcs/pull/388/commits/5c4d2e10a9e265210608339fcd95627d9c399f60)

**Comment by ktarbet** [comment]
The SHEF decoding is nearly working in the test.  However, the last value (of the SHEF time-series) is missing when using the Junit/DecodesHelper.  When using the UI the last value is OK.

**Comment by MikeNeilson** [comment]
It was probably to be a separator for the algorithms, but for the life of me I don't remember why i put it there.
And I like what you setup here so let's stick with your change and we'll figure out the conversion algorithms later.

**Comment by MikeNeilson** [comment]
The last value being missing is odd, the test is calling `decodesScript.decodeMessage` just as the GUI would. maybe add an extra newline to the input file?

**Comment by ktarbet** [comment]
Thanks- extra newline fixed the issue.

**Comment by MikeNeilson** [comment]
Givwen that fixed it that is in itself a bug. I dont think a blank newline
shpuld be required.

File: src/main/java/decodes/decoder/ShefProcess.java [hunk 1]
```diff
@@ -218,31 +218,30 @@ else if (c == '\n')
 				if (c == ':')
 				{
 					lineContext = LineContext.InComment;
-					debug(idx, "start of commnet");
-					continue;
+					debug(idx, "start of comment");
 				}
 				else if (Character.isWhitespace(c) || c == '/')
```
Comment (MikeNeilson) [inline src/main/java/decodes/decoder/ShefProcess.java:223]: This all looks reasonable but would probably be good to also include a simple test with the .A format to verify we aren't going to be adding a regression there.

Comment (ktarbet) [reply src/main/java/decodes/decoder/ShefProcess.java:223]: The header parsing for shef  (ShefPMParser.java) doesn't check for comment lines!

If I remove the comment lines from the sample .A data it appears to be happy.

File: src/main/java/decodes/datasource/ShefPMParser.java [hunk 2]
```diff
@@ -196,6 +147,56 @@ else if (dateField.length() == 8)
 		msg.setPM(GoesPMParser.SITE_NAME, new Variable(stationName));
```
[the remark this answers is not in the document]
Comment (ktarbet) [reply src/main/java/decodes/datasource/ShefPMParser.java:137]: Depending on the data set we have options.
We could subtract leading comment lines from the MESSAGE_LENGTH . 

For a simple .E format the header is a little more concrete.

Preferably, we have definition for header length setHeaderLength() , that is compatible with MESSAGE_LENGTH .

File: src/main/java/decodes/datasource/ShefPMParser.java [hunk 3]
```diff
@@ -191,11 +133,85 @@ else if (dateField.length() == 8)
...
 		Logger.instance().debug3("SHEF Message Date=" + ccyymmdd.format(msgTime));
 		msg.setPM(PM_TIMEZONE, new Variable(tz.getID()));
-		msg.setPM(GoesPMParser.MESSAGE_LENGTH, new Variable(data.length - headerEnd));
+		msg.setPM(GoesPMParser.MESSAGE_LENGTH, new Variable(data.length));
 		msg.setPM(GoesPMParser.FAILURE_CODE, new Variable('G'));
 		msg.setPM(GoesPMParser.SITE_NAME, new Variable(stationName));
 	}
 
+
+	private Date parseDate(String dateField, TimeZone tz) throws Exception {
```
Comment (MikeNeilson) [inline src/main/java/decodes/datasource/ShefPMParser.java:142]: Sorry, I didn't catch this before. the opening `{` needs to be on the next line.

File: src/test/java/fixtures/DecodesHelper.java [hunk 4]
```diff
@@ -112,6 +115,15 @@ public static Arguments getScript(String testName, String resourcePath) throws D
                 decodesScript.scriptSensors.add(stage);
                 ConfigSensor configSensor = new ConfigSensor(decodesScript.platformConfig, Integer.parseInt(parts[0]));
                 configSensor.sensorName = parts[1];
+
+                if( parts.length >= 5) {
```
Comment (MikeNeilson) [inline src/test/java/fixtures/DecodesHelper.java:119]: `{` on next line.

File: src/test/java/fixtures/DecodesHelper.java [hunk 5]
```diff
@@ -292,17 +312,41 @@ public static DecodedSample sampleFor(int sensor, ZonedDateTime sampleTime,
...
         }
-
+        String info = samplesToString(sensor, samples);
         throw new NoSuchElementException("Expected Sample for sensor " 
                                      + sensor + " at time "
-                                     + sampleTime + " doesn't exist");
+                                     + sampleTime + " doesn't exist\nDecoded "+samples.size()+" samples\n"+info);
+    }
+
+    private static String samplesToString(int sensor, ArrayList<DecodedSample> samples) {
```
Comment (MikeNeilson) [inline src/test/java/fixtures/DecodesHelper.java:328]: `{` on next line

**Review by MikeNeilson — CHANGES_REQUESTED** [review]
Minor formatting changes required. Otherwise the logic of this looks great to me.

**Review by MikeNeilson — APPROVED** [review]

**Pull request merged**